packages/plugins/response-cache/src/plugin.ts

```diff
@@ -206,38 +209,40 @@ export type ResponseCacheExecutionResult = ExecutionResult<
 >;
 
 const originalDocumentMap = new WeakMap<DocumentNode, DocumentNode>();
-const addTypeNameToDocument = memoize1(function addTypeNameToDocument(
+const addTypeNameToDocument = memoize2(function addTypeNameToDocument(
   document: DocumentNode,
+  addTypeNameToDocumentOpts: { invalidateViaMutation: boolean; invalidateViaSubscription: boolean },
 ): DocumentNode {
-  let documentChanged = false;
   const newDocument = visit(document, {
-    SelectionSet(node): SelectionSetNode {
-      if (
-        !node.selections.some(
-          selection => selection.kind === Kind.FIELD && selection.name.value === '__typename',
-        )
-      ) {
-        documentChanged = true;
-        return {
-          ...node,
-          selections: [
-            {
-              kind: Kind.FIELD,
-              name: {
-                kind: Kind.NAME,
-                value: '__typename',
-              },
-            },
-            ...node.selections,
-          ],
-        };
+    SelectionSet(node, _key, parent) {
+      if (parent && 'kind' in parent && parent.kind === Kind.OPERATION_DEFINITION) {
+        if (parent.operation === 'mutation' && !addTypeNameToDocumentOpts.invalidateViaMutation) {
+          return BREAK;
```

Comment:
You probably want to return `false` here ? If the mutation invalidation is disabled, we can skip the entire subtree of the AST

Reply:
Good catch! Thanks!